### PR TITLE
Include commit SHA in `[AssemblyInformationalVersion]` value

### PR DESCRIPTION
**Pull request by dougbu**
- see dotnet/arcade#5866 discussion

### CODE REVIEW

**Review by wtgodbe — APPROVED** [review]

**Comment by dougbu** [comment]
/fyi @ajcvickers @bricelam @smitpatel 

@Pilchie this is the other 5.0 fix to use consistent "format" for our `[AssemblyInformationalVersion]` attributes. Will do 3.1.10 variants after checking 3.1.9 assemblies. Suspect I'll need to change dotnet/extensions too.

**Comment by ajcvickers** [comment]
Note that we do expose and use `[AssemblyInformationalVersion]` in model snapshots, etc. I doubt this will break anything, since I don't think we do any parsing. @bricelam?

**Comment by ajcvickers** [comment]
Ref: https://github.com/dotnet/efcore/blob/main/src/EFCore/Infrastructure/ProductInfo.cs

**Comment by bricelam** [comment]
It breaks—the column length is too small in the database

**Comment by bricelam** [comment]
Well, it broke when we tried this in the past anyway

**Comment by ajcvickers** [comment]
If only we had a way to update the schema... :trollface: ;-)

**Comment by dougbu** [comment]
> Well, it broke when we tried this in the past anyway

Do tests cover any database mapping from the `ProductInfo` class❔ PR validation succeeded.

**Comment by bricelam** [comment]
I think it was on MySQL or Oracle or something. We should review the previous attempt before moving forward on this.

**Comment by bricelam** [comment]
Migrating the history table might be a bit too chicken and egg. 😐

**Comment by dougbu** [comment]
Where's the previous attempt❔ Would like to get this into 5.0 RC2. We have a bit longer for the 3.1 version of this (#22650).

**Comment by dougbu** [comment]
Found previous attempt I think. @bricelam [your comment](https://github.com/dotnet/efcore/pull/14177#issuecomment-448413563) in #14177 (the original Arcade conversion you did) says
> Fixed the version, broke some tests.

> ``` csharp
> [AssemblyInformationalVersion("3.0.0-preview.18618.5+078a3054c6c04951dd00c71c6d3ad85d4cb350d6")]
> ```

> But `__EFMigrationsHistory.ProductVersion` only allows up to 32 characters.

> We can remove the commit part by setting `IncludeSourceRevisionInInformationalVersion` to `false`.

So, what controls the length of `__EFMigrationsHistory.ProductVersion` and why aren't tests failing here and in #22650❔

**Comment by ajcvickers** [comment]
@dougbu We likely need to add coverage.

Is there a functional reason for making this change?

**Comment by dougbu** [comment]
> Is there a functional reason for making this change?

Well, perhaps a corner case, but not including the SHA makes some Reflection introspection difficult or, in the EF case, impossible.

@sebastienros pointed out inconsistent assembly metadata between aspnetcore and runtime DLLs in dotnet/arcade#5866. That led us to look at other repos and their assemblies and we noticed EF assemblies don't contain the commit SHA _anywhere_. We also saw [Arcade documentation](https://github.com/dotnet/arcade/blob/master/Documentation/CorePackages/Versioning.md#assembly-informational-version-generation) stressing use of this "full" `[AssemblyInformationalVersion]` format.

If we find including the SHA in the `[AssemblyInformationalVersion]` causes problems, I suggest adding `[AssemblyMetadata("CommitHash", ...)]`. That's an option aspnetcore uses though it's now redundant.

Realize the PR validation build doesn't exercise this change; `[AssemblyInformationalVersion]` values remain `5.0.0-ci` in an aspnetcore validation build that includes a similar update. I'll kick off an internal build to (a) ensure we get artifacts to double-check and (b) run the tests w/ values like `5.0.0-rc.2.20472.5+93f73c77640abc472fb901e18bf5055961e23e44`.

**Comment by bricelam** [comment]
+1 for using `[AssemblyMetadata("CommitHash", ...)]` instead

**Comment by bricelam** [comment]
Here's [the comment](https://github.com/dotnet/efcore/pull/14177#issuecomment-448413563) from last time we tried. The column length is 32.
```
                                v
3.0.0-preview.18618.5+078a3054c6c04951dd00c71c6d3ad85d4cb350d6
                                ^
```

**Comment by ajcvickers** [comment]
@bricelam Could we update ProductInfo.cs to trim the value coming from the attribute?

**Comment by dougbu** [comment]
@bricelam [that comment](https://github.com/dotnet/efcore/pull/14177#issuecomment-448413563) looks like a separate signing issue hit in #14177 that @JohnTortugo fixed in Arcade. But the
```
                                v
3.0.0-preview.18618.5+078a3054c6c04951dd00c71c6d3ad85d4cb350d6
                                ^
```
bit is definitely relevant.

I think https://github.com/dotnet/efcore/pull/14177#issuecomment-448413563 remains the most relevant comment in #14177.

Not sure why https://dev.azure.com/dnceng/internal/_build/results?buildId=825609 isn't failing (yet❔) though.

**Comment by bricelam** [comment]
@ajcvickers I'd want to add [a full semantic version parser](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/NuGet.Versioning) as part of that.

**Comment by ajcvickers** [comment]
@bricelam Consider for 6.0?

**Comment by bricelam** [comment]
Why not just use an AssemblyMetadata attribute instead? And do it in RC2.

**Comment by ajcvickers** [comment]
@dougbu 

> Well, perhaps a corner case, but not including the SHA makes some Reflection introspection difficult or, in the EF case, impossible.

Can you link to more information about this? What, specifically, is impossible?

**Comment by ajcvickers** [comment]
@bricelam I'm still trying to figure out the actual reason for this change so that if we do that we know whether or not it fixes anything.

**Comment by bricelam** [comment]
Additional context: We lost this information when we moved form KoreBuild to Arcade. I thought I filed an issue to track adding it back, but it must have fell through the cracks.

**Comment by dougbu** [comment]
> Can you link to more information about this?

See dotnet/arcade#5866 and https://github.com/dotnet/arcade/blob/master/Documentation/CorePackages/Versioning.md#assembly-informational-version-generation

> What, specifically, is impossible?

Figuring out the commit SHA used when building an assembly.

> We lost this information when we moved form KoreBuild to Arcade

and

> +1 for using `[AssemblyMetadata("CommitHash", ...)]` instead

I'm fine with this because https://dev.azure.com/dnceng/internal/_build/results?buildId=825609 was **successful** despite assemblies having `[AssemblyInformationalVersion("5.0.0-rc.2.20472.2+fba9fd552018bb9bde9e6424c1a7c3db97721265")]`. Would make the metadata available though it wouldn't match the Arcade guidance or @sebastienros's hope for consistency.

I agree this change would need to be larger i.e. include tests and either a column width update or a parser in `ProductInfo`. Either is risky this late in 5.0 and not reasonable at all for 3.1.

Any objection to making the minimal `[AssemblyMetadata("CommitHash", ...)]` addition and leaving `[AssemblyInformationalVersion(...)]` alone❔

**Comment by ajcvickers** [comment]
@dougbu @bricelam Thanks for the detailed info. Seems like the CommitHash approach is the way to go.

File: Directory.Build.targets [hunk 1]
```diff
@@ -1,3 +1,15 @@
 <Project>
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.Arcade.Sdk" />
+
+  <Target Name="GetCustomAssemblyAttributes"
+          BeforeTargets="GetAssemblyAttributes"
+          Condition=" '$(MSBuildProjectExtension)' == '.csproj' "
```
Comment (bricelam) [inline Directory.Build.targets:6]: Why this condition?

**Review by bricelam — APPROVED** [review]

**Comment by dougbu** [comment]
:up::date: to switch approaches

dotnet/aspnetcore also adds the following in its version of the same target.
``` msbuild
<AssemblyAttribute Include="System.Reflection.AssemblyMetadataAttribute" Condition="$(RepositoryUrl.StartsWith('https://github.com'))">
  <_Parameter1>SourceCommitUrl</_Parameter1>
  <_Parameter2>$(RepositoryUrl)/tree/$(SourceRevisionId)</_Parameter2>
</AssemblyAttribute>

<AssemblyAttribute Include="System.Reflection.AssemblyMetadataAttribute" Condition="'$(Serviceable)' == 'true'">
  <_Parameter1>Serviceable</_Parameter1>
  <_Parameter2>True</_Parameter2>
</AssemblyAttribute>
```
`SourceCommitUrl` information looks a bit redundant since assemblies already contain `[AssemblyMetadata("RepositoryUrl", "https://github.com/dotnet/efcore")]` and will now contain `[AssemblyMetadata("CommitHash", ...)]`. Not sure anything would read the `SourceCommitUrl` data.

But, is the lack of `Serviceable` metadata a problem @blowdart❔

File: Directory.Build.targets (hunk 1 above)
Comment (bricelam) [reply Directory.Build.targets:6]: Well, I suppose we do have `helix.proj`...

Comment (dougbu) [reply Directory.Build.targets:6]: I didn't check whether this repo contains anything else but C# projects and copied the condition from dotnet/aspnetcore. Let me know if it's useless.

**Comment by bricelam** [comment]
> But, is the lack of Serviceable metadata a problem

IIRC, it was only used on .NET Framework. See https://github.com/dotnet/arcade/issues/1526

**Comment by blowdart** [comment]
> IIUC, it was only used on .NET Framework. See [dotnet/arcade#1526](https://github.com/dotnet/arcade/issues/1526)

I don't believe that's true. The Servicable attribute was supposed to stay in place to enable the runtime to drop information on what assemblies have ever been loaded on a system, and that includes on Core. I don't believe this has ever changed.

Put it back, unless @ericstj says hammer patching uses another mechanism.

**Comment by dougbu** [comment]
> Put it back, unless @ericstj says hammer patching uses another mechanism.

Easy to do though the `Serviceable` information hasn't been included in EF attributes at least since the migration to Arcade. I'll give @ericstj a chance to respond.

**Comment by ericstj** [comment]
Serviceable assembly metadata is used by .NETFramework loader to write breadcrumbs.  I believe the .NETCore host uses the serviceable metadata from the deps file.  cc @agocke @vitek-karas

**Comment by dougbu** [comment]
Hmm, I believe this repo produces a number of assemblies usable on .NET Framework e.g. Micrsoft.Data.Sqlite as seen at https://github.com/dotnet/efcore/blob/b970bf29a46521f40862a01db9e276e6448d3cb0/src/Microsoft.Data.Sqlite/Microsoft.Data.Sqlite.csproj#L18

**Comment by bricelam** [comment]
The nuspecs include `<serviceable>true</serviceable>`, which is maybe what .NET Core uses instead.

**Comment by bricelam** [comment]
Happy to add it back--that's why I filed the issue on Aracade. 🙂 (to get to the bottom of whether it was actually needed or not)

**Comment by dougbu** [comment]
I need lunch but will at least add the `Serviceable` information after that. Not hearing a clamor for `SourceCommitUrl` but let me know if you want that too.

Will get this ready to go and then update the 3.1 PR.

Any reason to worry about EF6's release/6.4 and master branches given how quiet that repo is❔

**Comment by dougbu** [comment]
@ajcvickers is this PR still blocked❔

**Comment by ajcvickers** [comment]
@dougbu Nope. @bricelam signing off is good for me. :-)

**Comment by bricelam** [comment]
Thanks, Doug! I'm really glad we have this information back. Reverse engineering it from the build number was painful the few times I had to do it.

**Pull request merged**